File: cpp/painthouse.cpp
class Solution {
public:
    int minCost(vector<vector<int>>& costs) {
        if (costs.empty()) return 0;
        vector<vector<int>> dp(costs.size(), vector<int>(3, 0));
        dp[0] = costs[0];
        for (int i = 1; i < costs.size(); i++) {
            for (int j = 0; j < costs[0].size(); j++) {
                dp[i][j] = costs[i][j]+min(dp[i-1][(j+1)%3], dp[i-1][(j+2)%3]);
            }
        }
        return min(dp.back()[0], min(dp.back()[1], dp.back()[2]));
    }
};

File: java/PaintHouse.java
class Solution {
    public int minCost(int[][] costs) {
        
        if(costs == null || costs.length == 0){
            return 0;
        }
        
        for (int i = 1; i < costs.length; i++) {
            for (int j = 0; j < 3; j++) {
                costs[i][j] += Math.min(costs[i-1][(j+1)%3], costs[i-1][(j+2)%3]);
            }
        }
        
        int n = costs.length-1;
        return Math.min(Math.min(costs[n][0], costs[n][1]), costs[n][2]);
    }
}
